Probe motor implementation (#47)
HEATRobotics/EMBR-Bot: Documentation/2025/probe-motor.md

# Rover Probe Motor Control System

# Overview
This documentation provides setup guide and serves as a resource to understand the implementation of the LMDCE421(probe-motor) communication.

This documentation hopes to achieve 2 goals:
1. Educate the reader with just enough knowledge to make incremental changes to the probe-motor codebase.

> 2. Educate the reader with enough knowledge to rebuild the probe-motor codebase.

Reviewer: ```suggestion
This document will give the reader sufficient knowledge to modify/rebuild the probe-motor codebase.
```

Reviewer: Suggesting this as original lines seem to be saying more or less the same goal

Author: I agree! Concise too.

The LMDCE421 offers an ethernet and serial interface. We have used the ethernet interface communication because there is extensive library support for encoding/decoding data easily over this interface and communication can easily be reproduced from any computer with an ethernet port. **MODBUS/TCP** is the chosen communication protocol for this interface.

## MODBUS/TCP
[MODBUS/TCP](https://www.novantaims.com/application-note/modbustcp-vs-mcodetcp-overview/) is a variant of the Modbus family of vendor-neutral communication protocols intended for supervision and control of automation equipment. 

<img width="646" height="259" alt="image" src="https://github.com/user-attachments/assets/230d1731-48d0-426b-b00b-496c96136636" />

* **Structure:** It wraps a standard Modbus frame inside a TCP/IP packet (Port 502).
* **Client/Server Model:** In this setup, our control script acts as the **Client** (Master), and the LMDCE421 acts as the **Server** (Slave).
* **Protocol Data Unit (PDU):** This is the core of the message containing the instruction.

<img alt="image" src="https://www.novantaims.com/support/mdi_getting_started/ethernet/images/pdu.gif" />

### Protocol Data Unit Examples

> The following table illustrates how the request parameters are structured within the PDU for common motor operations:

Reviewer: Noticed in _write_registers you use Little-endian. May be worth having documented assuming that whatever reads from same address will need to handle bytes in the same fashion.

Author: Great idea. Included it for [write_registers](https://github.com/HEATRobotics/EMBR-Bot/commit/4bdb62691f9d8b670619944df64759d447d37dbd) and [read_holding_registers](https://github.com/HEATRobotics/EMBR-Bot/commit/1a6cbe5e373a5924a01be42f47796403d60a90a1)

| Operation | Function Code | Data Address | Data Quantity/Value | Observation |
| :--- | :--- | :--- | :--- | :--- |
| **Read Position** | `03` (Read Holding) | `00 64` (100) | `00 02` | Reads two registers to get a 32-bit position value. |
| **Check Status** | `04` (Read Input) | `00 0A` (10) | `00 01` | Checks a single register for the "Ready" flag. |
| **Set Velocity** | `06` (Write Single) | `01 2C` (300) | `03 E8` | Writes a fixed speed (e.g., 1000) to the velocity register. |

> | **Multi-Move** | `10` (Write Multiple) | `02 58` (600) | `00 04` | Updates acceleration, deceleration, and target at once. |

Reviewer: ```suggestion
| **Multi-Move** | `10` (Write Multiple) | `02 58` (600) | `00 04` | Updates acceleration/deceleration, and target at once. |
```

```

### 2. write_register
Used to write a single 16-bit value to the motor.
```python
# Write the value 1 to register address 50
client.write_register(address=50, value=1)
```

### 3. write_registers
Used for 32-bit values or simultaneous parameter updates.
```python
# Write a list of values to registers starting at address 200
values = [1000, 500] 
client.write_registers(address=200, values=values)
```

### 4. read_input_registers
Used to read hardware-level data that is typically read-only.
```python
# Read the current position from the input register
pos_data = client.read_input_registers(address=300, count=1)
```
---

This is a hardware-abstracted ROS2 interface for controlling a linear probe motor (LMDCE421 Stepper). It features a **Factory Pattern** that allows for seamless switching between real hardware and simulated environments via a JSON configuration.

## ROS2 Network
![Text](Images/Probe%20Motor%20Implementation%20Overview.png)

The system is divided into three layers:

1. **Hardware/Simulation Layer**: Handles Modbus TCP communication or simulated physics.
2. **Abstraction Layer**: The `StepperFactory` uses `sensors.json` to decide which driver to instantiate.

> 3. **ROS2 Layer**: The `probeStepper` node acts as the server, while `nav` acts as the mission controller.

Reviewer: 'probeStepper' -> 'probeMotor'?

> 3. **ROS2 Layer**: The `probeStepper` node acts as the server, while `nav` acts as the mission controller.

Reviewer: ```suggestion
3. **ROS2 Layer**: The `probeMotor` node acts as the server, while `nav` acts as the mission controller.
```

| Parameter | Description |
| --- | --- |
| `mode` | Set to `"real"` for hardware or `"sim"` for simulation. |
| `steps_per_distance` | Conversion factor (Steps per cm). Default: `51200`. |
| `motion_range` | Maximum vertical displacement allowed (cm). |
| `max_velocity` | Speed limit in steps/second. |

---

### 3. ROS2 Interface

#### Subscribed Topics

* **`/move_probe_motor`** (`msg_interface/MoveProbeMotor`): Receives target position commands (cm).
* **`/probe_motor_feedback`** (`msg_interface/ProbeMotorFeedback`): Receives position updates and movement confirmation.

#### Message Definitions

* **MoveProbeMotor**: `int32 move_position` (Target depth in cm).
* **ProbeMotorFeedback**: `bool has_moved`, `int32 position` (Current depth in cm).

---

### 4. Setup and Execution

#### Building the Package

```bash
cd ~/ros2_ws
colcon build --packages-select embr msg_interface
source install/setup.bash

```

#### Running the System

To launch all nodes (Temperature, Cube, Radio, and Motor) with the default configuration:

```bash
ros2 launch embr embr_launch.py

```

To specify a custom configuration file:

```bash

> ros2 launch embr embr_launch.py config_file:=/path/to/your_config.json

Reviewer: Need to update DockerFile to include pymodbus or in documentation mention mandatory pymodbus install via pip

Author: Added it to the dependencies in Tools/Setup-Scripts/[install-dependencies](https://github.com/HEATRobotics/EMBR-Bot/commit/4921ba3b4ab455b75d6167780c6ea22456d688b3)
